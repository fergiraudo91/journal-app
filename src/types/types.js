export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetErrors: '[UI] Set Errors',
    uiRemoveErrors: '[UI] Remove Errors',

    uiStartingLoading: '[UI] Starting Loading',
    uiFinishLoading: '[UI] Finish Loading',

    notesAddNew: '[Notes] New Note',
    notesActive: '[Notes] Set Active Note',
    notesLoad: '[Notes] Notes Load',
    notesUpdated: '[Notes] Update Notes',
    notesFileURL: '[Notes] Updated Image URL',
    notesDelete: '[Notes] delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning'
}