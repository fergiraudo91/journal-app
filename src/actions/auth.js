import Swal from 'sweetalert2';
import { types } from "../types/types";
import { firebase, googleAddProvider } from "../firebase/firebase-config";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((e) => {
        Swal.fire('Error', e.message, 'error');
        dispatch(finishLoading());
      });
  };
};

export const startRegisterEmailPasswordName = (name, email, password) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((e) => {
        Swal.fire('Error', e.message, 'error');
      });
  };
};

export const googleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAddProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const login = (uid, displayName) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  };
};
