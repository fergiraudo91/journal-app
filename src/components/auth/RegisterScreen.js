import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";
import { startRegisterEmailPasswordName } from "../../actions/auth";
import { removeError, setError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";


export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { msError } = useSelector(state => state.ui);

  const [inputValues, handleInputChange] = useForm({
    name: "",
    email: "",
    password: "",
    password2: "",
  });
  const { name, email, password, password2 } = inputValues;

  const handleRegister = (e) => {
    e.preventDefault();
    console.log(name, email, password, password2);
    if (isFormValid()) {
      dispatch(startRegisterEmailPasswordName(name, email, password));
    }
  };

  const isFormValid = () => {
    if (name.trim().length <= 2) {
      dispatch(setError("Name is required"));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError("Email is not valid"));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          "Password should be at least 6 characters long and match each other"
        )
      );
      return false;
    }

    dispatch(removeError());

    return true;
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>

      <form 
      className="animate__animated animate__fadeIn"
      onSubmit={handleRegister}>
        {

         msError && <div className="auth__alert-error">{msError}</div>
        }
        <input
          type="text"
          autoComplete="off"
          placeholder="Name"
          name="name"
          className="auth__input"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          autoComplete="off"
          placeholder="Email"
          name="email"
          className="auth__input"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="*******"
          name="password"
          className="auth__input"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Confirm"
          name="password2"
          className="auth__input"
          value={password2}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary btn-block mb-5">
          Register
        </button>

        <Link className="link" to="/auth/register mt-5">
          Alredy Register?
        </Link>
      </form>
    </>
  );
};
