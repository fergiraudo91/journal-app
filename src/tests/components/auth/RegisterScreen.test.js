import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import React from 'react';
import { Provider } from 'react-redux';
import thunk from "redux-thunk";
import configureStore from 'redux-mock-store' //ES6 modules
import { mount } from "enzyme";
import { MemoryRouter } from 'react-router';
import { types } from "../../../types/types";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msError: null
    }
}

let store = mockStore(initState);

describe('Pruebas en <RegisterScreen/>', () => {

    const wrapper = mount(
      
        <Provider store  = {store}>
            <MemoryRouter>
                <RegisterScreen/>
            </MemoryRouter>
        </Provider>
        
    );
   
    test('Debe mostrarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot();

    });
    
    test('Debe de hacer el dispatch de la accion respectiva ', () => {
       
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', {
            target: '',
            name: 'email'
        });

        wrapper.find('form').simulate('submit');

        const actions = store.getActions();

        expect(actions[0]).toEqual({
            type: types.uiSetErrors,
            payload: 'Name is required'
        });

    });

    test('Debe de mostrar la caja de alerta con el error', () => {

        const initState = {
            auth: {},
            ui: {
                loading: false,
                msError: 'Email no es correcto'
            }
        }

        let store = mockStore(initState);

        const wrapper = mount(
      
            <Provider store  = {store}>
                <MemoryRouter>
                    <RegisterScreen/>
                </MemoryRouter>
            </Provider>
        );


        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msError);
        
    })
    
    
    
});
