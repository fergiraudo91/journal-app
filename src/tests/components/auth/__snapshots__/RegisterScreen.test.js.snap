// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen/> Debe mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="*******"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/register mt-5"
      onClick={[Function]}
    >
      Alredy Register?
    </a>
  </form>,
]
`;
